uber/jinja.py

```diff
@@ -1,4 +1,84 @@
 from uber.common import *
+from functools import lru_cache
+from jinja2.loaders import split_template_path
+from jinja2.utils import open_if_exists
+from jinja2.exceptions import TemplateNotFound
+
+
+class MultiPathEnvironment(jinja2.Environment):
+
+    @request_cached_property
+    def _templates_loaded_for_current_request(self):
+        return set()
+
+    @lru_cache()
+    def _get_matching_filenames(self, template):
+        pieces = split_template_path(template)
+        matching_filenames = []
+        for searchpath in self.loader.searchpath:
+            filename = os.path.join(searchpath, *pieces)
+            if os.path.exists(filename):
+                matching_filenames.append(filename)
+        return matching_filenames
+
+    def _load_template(self, name, globals):
+        """
+        Overridden to consider templates already loaded during the current request.
+        """
+        if self.loader is None:
+            raise TypeError('no loader for this environment specified')
+
+        loaded_templates = self._templates_loaded_for_current_request
```

Comment:
Random style nitpick: you declare ``loaded_templates`` here but then only use it once and later in this method you revert to just referencing ``self._templates_loaded_for_current_request`` directly.  You could probably just refer to ``loaded_templates`` throughout the function for both brevity and consistency.

Reply:
Yeah, so I declared `loaded_templates` just to keep line-length down. But then later on I explicitly use `self._templates_loaded_for_current_request` to make clear that we are updating a persistent attribute of the object.

Here's my thinking, let me know how you feel about it. Even though `loaded_templates` and `self._templates_loaded_for_current_request` point to the same object in memory, if a new developer were reading the source code they would have to understand that context. So, it seemed like this:
```python
template = self.loader.load(self, filename, globals)
self._templates_loaded_for_current_request.add(template.filename)
```
was more obvious than this:
```python
template = self.loader.load(self, filename, globals)
loaded_templates.add(template.filename)
```

Because to truly understand what that line of code is doing, you must look up the declaration of `loaded_templates`. So maybe it would be better to not declare `loaded_templates` at all?

Perhaps I made the variable name for `_templates_loaded_for_current_request` way too long? I wanted to make it very obvious exactly what that property represented.

I don't know!

Reply:
That is indeed a tough "naming things" question.  I initially thought about suggesting that you rename ``self._templates_loaded_for_current_request`` to just be ``self._loaded_templates``, but then I decided against making that suggestion due to the fact that it might mislead someone into assuming that the templates were only loaded once period instead of per request.

I'm actually kind of a fan in some cases like this of having shorter and more cryptic names for private fields.  For example, if we named this ``self._rtc`` and then in the docstring said something like "rtc stands for 'request template cache', as this is the cache of templates which have been loaded for the currently-being-served request" then it would be short enough to keep the line length down without being misleading. 

One often-overlooked benefit of cryptic variable names like ``_rtc`` is that a newcomer seeing it for the first time won't read the variable name and immediately come to any misconceptions about what it is, because they *must* read the docstring to have any idea what it is!

This is obviously not a universally accepted or popular approach to naming, and it's not one that I would take for public fields on a class, but I'm personally a fan of doing it for private fields and methods.  Having longer variable names is more common (at my first job there was a variable named ``bulk_mail2_a0_width_increase_for_pass_trough_percent``), so it's mostly a style preference.